### PR TITLE
Add web API to delete/disable environment

### PR DESCRIPTION
**Pull request by nakabonne**
**What this PR does / why we need it**:
For applications that belong to the deleted/disabled environment, let me discuss on https://github.com/pipe-cd/pipe/issues/2017.

**Which issue(s) this PR fixes**:

Fixes https://github.com/pipe-cd/pipe/issues/1997

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by nakabonne** [comment]
Have a look at the modified part of TypeScript.
/cc @cakecatz

**Comment by nakabonne** [comment]
the web test looks fail, it went well on my laptop though...

**Comment by cakecatz** [comment]
Nice!
/lgtm

File: pkg/app/api/grpcapi/web_api.go [hunk 1]
```diff
@@ -168,6 +170,107 @@ func (a *WebAPI) ListEnvironments(ctx context.Context, req *webservice.ListEnvir
...
+	if enable {
+		updater = a.environmentStore.EnableEnvironment
+	} else {
+		updater = a.environmentStore.DisableEnvironment
+	}
+
+	if err := updater(ctx, envID); err != nil {
+		switch err {
+		case datastore.ErrNotFound:
+			return status.Error(codes.InvalidArgument, "The environment is not found")
```
Comment (khanhtc1202) [inline pkg/app/api/grpcapi/web_api.go:237]: should be `codes.NotFound` 👀

File: pkg/app/api/grpcapi/web_api.go [hunk 2]
```diff
@@ -168,6 +170,107 @@ func (a *WebAPI) ListEnvironments(ctx context.Context, req *webservice.ListEnvir
...
+
+	env, err := getEnvironment(ctx, a.environmentStore, envID, a.logger)
+	if err != nil {
+		return err
+	}
+	a.envProjectCache.Put(envID, env.ProjectId)
+
+	if env.ProjectId != projectID {
+		return status.Error(codes.PermissionDenied, "Requested environment doesn't belong to the project you logged in")
+	}
```
Comment (khanhtc1202) [inline pkg/app/api/grpcapi/web_api.go:270]: should we validate by L268 if condition before update the cache 🤔

File: pkg/app/api/grpcapi/web_api.go (hunk 1 above)
Comment (nakabonne) [reply pkg/app/api/grpcapi/web_api.go:237]: It should. All parts including Applications should get so.

File: pkg/app/api/grpcapi/web_api.go (hunk 2 above)
Comment (nakabonne) [reply pkg/app/api/grpcapi/web_api.go:270]: I would say no. `envID` and `env.ProjectId` at L266 are exactly the correct pair.
The validation target at this function is `projectID`. Why you got confused may be the If condition order; we should make the order as:
```go
if projectID != env.ProjectId
```

Comment (khanhtc1202) [reply pkg/app/api/grpcapi/web_api.go:270]: oh, I get your point now 👍  Thanks for clarifying 🙏

File: pkg/app/api/grpcapi/web_api.go (hunk 1 above)
Comment (nakabonne) [reply pkg/app/api/grpcapi/web_api.go]: Addressed 651e3a8

**Comment by khanhtc1202** [comment]
Thank you 🙏 
/lgtm

File: pkg/app/api/grpcapi/web_api.go [hunk 3]
```diff
@@ -168,6 +170,107 @@ func (a *WebAPI) ListEnvironments(ctx context.Context, req *webservice.ListEnvir
...
+}
+
+func (a *WebAPI) DisableEnvironment(ctx context.Context, req *webservice.DisableEnvironmentRequest) (*webservice.DisableEnvironmentResponse, error) {
+	if err := a.updateEnvironmentEnable(ctx, req.EnvironmentId, false); err != nil {
+		return nil, err
+	}
+	return &webservice.DisableEnvironmentResponse{}, nil
+}
+
+func (a *WebAPI) DeleteEnvironment(ctx context.Context, req *webservice.DeleteEnvironmentRequest) (*webservice.DeleteEnvironmentResponse, error) {
```
Comment (nghialv) [inline pkg/app/api/grpcapi/web_api.go:187]: Didn't we decide that it should be checked to ensure no piped is using the env before allowing disabling/deleting it?
I'm not sure if I remember correctly.

Comment (nakabonne) [reply pkg/app/api/grpcapi/web_api.go:187]: We did. I'm looking to discuss that on https://github.com/pipe-cd/pipe/issues/2017 as mentioned in the PR body.

Comment (nghialv) [reply pkg/app/api/grpcapi/web_api.go:187]: Nice. Let's go.

**Pull request merged**